### ISSUE TITLE
Cannot connect ws client with asynchWebServer Socket

### ISSUE THREAD

**Issue by MoegerleStephan** [Links2004/arduinoWebSockets#159]
Hi there,
I've investigate couples of hours to connect your asynch WebSocket (ws) client with the integrated ws server of the async webserver (https://github.com/me-no-dev/ESPAsyncWebServer/tree/master/src).
**-> How is this possible?** 
I'm planing to use one ws Server for communication between Website and Master ESP as well as for the communication between a second ESP and the Master ESP.
Therefore I need to implement your ws Client on the second ESP, that communicates with the Master ESP.

The two Sketches that I'm using for tests are: 
- WebSocketClient.ino (arduinoWebSockets)
- ESP_AsyncFSBrowser.ino (ESPAsyncWebServer)

On webpage side I use following command what works fine for me
  `websock = new WebSocket('ws://'+document.location.host+'/ws',['arduino']);`

For heap reasons I want to avoid another WebSocket Server on another port on the Master ESP.
Therefore I've tried following command on the WebSocketClient.ino:

  `webSocket.begin("ws://192.168.178.125/ws", 80);`

The result was:
> [WS-Client] asyncConnect...
> [WS-Client] connection to ws://192.168.178.125/ws:80 Faild

Another try was establishing a connection on the 81 port. The result was following:
> [WS-Client] connection to ws://192.168.178.125/ws:81 Faild
> [WS-Client] asyncConnect...

Last try was with following command:
  `webSocket.begin("192.168.178.125", 80);`

The result was:
> [WS-Client][sendHeader] sending header... Done (17818us).
> [WS-Client][handleHeader] RX: HTTP/1.1 200 OK
> [WS-Client][handleHeader] RX: Content-Length: 2504
> [WS-Client][handleHeader] RX: Content-Type: text/html
> [WS-Client][handleHeader] RX: Content-Disposition: inline; filename="index.htm"
> [WS-Client][handleHeader] RX: Connection: close
> [WS-Client][handleHeader] RX: Accept-Ranges: none
> [WS-Client][handleHeader] Header read fin.
> [WS-Client][handleHeader] Client settings:
> [WS-Client][handleHeader]  - cURL: /
> [WS-Client][handleHeader]  - cKey: zwTEhLVUtnm1j3H0S67QjA==
> [WS-Client][handleHeader] Server header:
> [WS-Client][handleHeader]  - cCode: 200
> [WS-Client][handleHeader]  - cIsUpgrade: 0
> [WS-Client][handleHeader]  - cIsWebsocket: 0
> [WS-Client][handleHeader]  - cAccept: 
> [WS-Client][handleHeader]  - cProtocol: 
> [WS-Client][handleHeader]  - cExtensions: 
> [WS-Client][handleHeader]  - cVersion: 0
> [WS-Client][handleHeader]  - cSessionId: 
> [WS-Client][handleHeader] no Websocket connection close.
> [WS-Client] client disconnected.
> [WSc] Disconnected!
> [WS-Server][0] Disconnect client
> [WS-Client] asyncConnect...
> [WS-Client] connected to 192.168.178.125:80.
> [WS-Client][sendHeader] sending header...
> [WS-Client][sendHeader] handshake GET / HTTP/1.1
> Connection: Upgrade
> Upgrade: websocket
> Sec-WebSocket-Version: 13
> Sec-WebSocket-Key: QDv8wuo9nzeRgaEe2hFyfw==
> Host: 192.168.178.125:80
> Origin: file://
> User-Agent: arduino-WebSocket-Client

I don't know what else I should configure. I've also tried to set the path as "/ws" and the protocol to "['arduino']" or rather "arduino". 

What I'm doing wrong?

Thanks and best regards, Stephan

**Comment by Links2004** [comment]
the current log shows that the server answers with a `index.htm` and not with a websocket connection.
I have never used the ESPAsyncWebServer as websocket server.

can you post the debug log with:
```cpp
webSocket.begin("192.168.178.125", 80, "/ws", "arduino");
```

**Comment by MoegerleStephan** [comment]
Thanks for your help! Log shows 400 Bad Request.. 

> [WS-Client] asyncConnect...
> [WS-Client] connected to 192.168.178.125:80.
> [WS-Client][sendHeader] sending header...
> [WS-Client][sendHeader] handshake GET /ws HTTP/1.1
> Connection: Upgrade
> Upgrade: websocket
> Sec-WebSocket-Version: 13
> Sec-WebSocket-Key: O3d/EDx+ZlmPL5qfQu3Itg==
> Host: 192.168.178.125:80
> Origin: file://
> User-Agent: arduino-WebSocket-Client
> 
> [WS-Client][sendHeader] sending header... Done (17915us).
> [WS-Client][handleHeader] RX: HTTP/1.1 400 Bad Request
> [WS-Client][handleHeader] RX: Content-Length: 0
> [WS-Client][handleHeader] RX: Connection: close
> [WS-Client][handleHeader] RX: Accept-Ranges: none
> [WS-Client][handleHeader] Header read fin.
> [WS-Client][handleHeader] Client settings:
> [WS-Client][handleHeader]  - cURL: /ws
> [WS-Client][handleHeader]  - cKey: O3d/EDx+ZlmPL5qfQu3Itg==
> [WS-Client][handleHeader] Server header:
> [WS-Client][handleHeader]  - cCode: 400
> [WS-Client][handleHeader]  - cIsUpgrade: 0
> [WS-Client][handleHeader]  - cIsWebsocket: 0
> [WS-Client][handleHeader]  - cAccept: 
> [WS-Client][handleHeader]  - cProtocol: 
> [WS-Client][handleHeader]  - cExtensions: 
> [WS-Client][handleHeader]  - cVersion: 0
> [WS-Client][handleHeader]  - cSessionId: 
> [WS-Client][handleHeader] no Websocket connection close.
> [WS-Client] client disconnected.
> [WSc] Disconnected!
> [WS-Server][0] Disconnect client

**Comment by Links2004** [comment]
ok then we need to look at the server side.
can you get any logs from there?

**Comment by MoegerleStephan** [comment]
Ok I've made some debug print commands in the Server WebSocket file.. 
The problem was in the following line: 
https://github.com/me-no-dev/ESPAsyncWebServer/blob/master/src/AsyncWebSocket.cpp#L867

the WS_STR_VERSION in Header (Sec-WebSocket-Version) AND the WS_STR_KEY in Header (Sec-WebSocket-Key) are missing..

What do you think about this?

**Comment by Links2004** [comment]
strange, the log shows that the client is sending it.

Header send from client:
```
GET /ws HTTP/1.1
Connection: Upgrade
Upgrade: websocket
Sec-WebSocket-Version: 13
Sec-WebSocket-Key: O3d/EDx+ZlmPL5qfQu3Itg==
Host: 192.168.178.125:80
Origin: file://
User-Agent: arduino-WebSocket-Client
```

@me-no-dev have you any idea whats going on?

**Comment by MoegerleStephan** [comment]
Hmm.. it seems that the request contains only one header: `Origin: file://`

```
      if (request->method() == HTTP_GET)
          os_printf("GET");
      else if (request->method() == HTTP_POST)
          os_printf("POST");
      else if (request->method() == HTTP_DELETE)
          os_printf("DELETE");
      else if (request->method() == HTTP_PUT)
          os_printf("PUT");
      else if (request->method() == HTTP_PATCH)
          os_printf("PATCH");
      else if (request->method() == HTTP_HEAD)
          os_printf("HEAD");
      else if (request->method() == HTTP_OPTIONS)
          os_printf("OPTIONS");
      else
          os_printf("UNKNOWN");
      os_printf(" http://%s%s\n", request->host().c_str(), request->url().c_str());

      if (request->contentLength()) {
          os_printf("_CONTENT_TYPE: %s\n", request->contentType().c_str());
          os_printf("_CONTENT_LENGTH: %u\n", request->contentLength());
      }

      int headers = request->headers();
      int i;
      for (i = 0; i < headers; i++) {
          AsyncWebHeader* h = request->getHeader(i);
          os_printf("_HEADER[%s]: %s\n", h->name().c_str(), h->value().c_str());
      }

      int params = request->params();
      for (i = 0; i < params; i++) {
          AsyncWebParameter* p = request->getParam(i);
          if (p->isFile()) {
              os_printf("_FILE[%s]: %s, size: %u\n", p->name().c_str(), p->value().c_str(), p->size());
          } else if (p->isPost()) {
              os_printf("_POST[%s]: %s\n", p->name().c_str(), p->value().c_str());
          } else {
              os_printf("_GET[%s]: %s\n", p->name().c_str(), p->value().c_str());
          }
      }
```
If I add the request output code above in this method (https://github.com/me-no-dev/ESPAsyncWebServer/blob/master/src/AsyncWebSocket.cpp#L867), the result was following:

> GET http://192.168.4.1:80/ws
> _HEADER[Origin]: file://

**Comment by me-no-dev** [comment]
is the client Internet Explorer or Edge? MS are a bunch of assholes and they do not follow RFC when constructing requests. Host header has to be the first header and in their case it's not. Async Web Server uses the Host header for routing and decision which headers to collect and when it does not come on time, headers get lost.

**Comment by me-no-dev** [comment]
Sorry @Links2004 me thinking browsers when you are clearly discussing your WebSocket client... too early for my brain I guess :) I'm glad it gave you an idea of what is wrong

**Comment by Links2004** [comment]
github auto closed it...
@me-no-dev nop it is my client ;) have moved the host header.
have you a link to the RFC would like to read it, only found https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.23 but there is no requirement for the position. 

@MoegerleStephan can you check if this helps?

**Comment by me-no-dev** [comment]
@Links2004 I had a user research it recently but can not seem to find the issue and the link he posted.
I have been meaning to get this issue "fixed" somehow but have been otherwise preoccupied and never got the chance (AsyncWebServer is not running yet on ESP32). I did not get to decision the few times tat I looked at the code though. Basically I need to collect all headers before any routing can be done and before handler is even attached (the point where decision which headers to be collected is made)
